### PR TITLE
Implementing svc

### LINKED ISSUE #5: Implement support vector machines
Tutorials and origin
https://www.analyticsvidhya.com/blog/2017/09/understaing-support-vector-machine-example-code/
https://www.csie.ntu.edu.tw/~cjlin/libsvm/
https://www.svm-tutorial.com/

Code - https://github.com/arnaudsj/libsvm

### PR DESCRIPTION
**Pull request by markroxor**
* **Please check if the PR fulfills these requirements**
  - [x] The commit message follows our guidelines or the commits will be squashed into one.
  - [x] Tests for the changes have been added.
  - [x] Docs have been added / updated.

* **What kind of change does this PR introduce?** 
     Implementing support vector classifier

* **Does this PR introduce a breaking change?** 
     None
* **This PR fixes/implements the following -**
     - [x] Docstrings.
     - [x] the linear kernel
     - [x] the polynomial kernel
     - [x] the RBF (Gaussian) kernel
     - [ ] the string kernel
     - [x] fix test fails cause by pickle's inherited class error.
     - [x] soft margin
     - [ ] Multivariate SVM
     - [x] tests

* **Closing issues**
       Resolves #5

### CODE REVIEW

File: fromscratchtoml/models/svm/svc.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        for i in range(effective_lagrange_multiplier_indices.shape[0]):
+            if effective_lagrange_multiplier_indices[i] == 1:
+                ind.append(i)
+
+        self.support_vectors = ch.stack([x for multiplier, x in zip(lagrange_multipliers, X) if multiplier > eta])
+        self.support_vectors_y = ch.Tensor(y[effective_lagrange_multiplier_indices])
+        self.effective_lagrange_multipliers = lagrange_multipliers[effective_lagrange_multiplier_indices]
+
+        self.b = 0
+        # TODO find a better way to calculate indexwise kernel trick
```
Comment (dikshant2210) [inline fromscratchtoml/models/svm/svc.py]: you can use torch.nonzero() to find indices of all non zero values in a tensor.

**Comment by dikshant2210** [comment]
Code coverage is a problem?

File: fromscratchtoml/test/test_support_vector_machines.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        # before predict.
+        model = svm.SVC()
+        with self.assertRaises(ModelNotFittedError):
+            model.predict(self.X)
+
+    def test_model_sanity(self):
+        # test when y is a list of integers (as in torch's dataloader implementation) our
+        # model is still sane.
+        # TODO
+        pass
```
Comment (dikshant2210) [inline fromscratchtoml/test/test_support_vector_machines.py]: add test case for prediction check

Comment (markroxor) [reply fromscratchtoml/test/test_support_vector_machines.py]: it is in my TODO, check the TODOs in the beginning of the PR

**Comment by markroxor** [comment]
It is still a WIP. Ignore what is not implemented.

File: fromscratchtoml/models/svm/svc.py (hunk 1 above)
Comment (dikshant2210) [reply fromscratchtoml/models/svm/svc.py]: great

File: fromscratchtoml/models/svm/svc.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    --------
+    >>> X1 = Distribution.linear(pts=100, mean=[8, 10], covr=[[1.5, 1], [1, 1.5]])
+    >>> X2 = Distribution.linear(pts=100, mean=[9, 5], covr=[[1.5, 1], [1, 1.5]])
+    >>> Y1 = ch.ones(X1.size()[0])
+    >>> Y2 = -ch.ones(X2.size()[0])
+    >>> X = ch.cat((X1, X2))
+    >>> y = ch.cat((Y1, Y2))
+    >>> clf_lin = svm.SVC(kernel='linear')
+    >>> clf_lin.fit(X, y)
+    >>> _, projections = clf_lin.predict(X, return_projection=True)
```
Comment (markroxor) [inline fromscratchtoml/models/svm/svc.py]: _ -> predictions.

**Comment by markroxor** [comment]
Solves issue #5

File: fromscratchtoml/toolbox/kernels.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    Returns
+    -------
+    kernel_trick : torch.Tensor
+                   Returns the kernel trick to transform the data into infinite
+                   dimensions.
+
+    """
+
+    euclidean_dist = pow(ch.norm(x - y), 2)
+    return np.exp(- euclidean_dist / (2 * pow(gamma, 2)))
```
Comment (markroxor) [inline fromscratchtoml/toolbox/kernels.py]: not returning torch.Tensor

File: fromscratchtoml/toolbox/kernels.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    https://nlp.stanford.edu/IR-book/html/htmledition/nonlinear-svms-1.html
+
+    Parameters
+    ----------
+    x : torch.Tensor
+    y : torch.Tensor
+
+    Returns
+    -------
+    kernel_trick : torch.Tensor
```
Comment (markroxor) [inline fromscratchtoml/toolbox/kernels.py]: returns float

File: fromscratchtoml/models/svm/svc.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        X = ch.Tensor(X)
+        y = ch.Tensor(y)
+        self.n = y.size()[0]
+
+        # create a gram matrix by taking the outer product of y
+        gram_matrix_y = ch.ger(y, y)
+        K = self.create_kernel_matrix(X)
+        gram_matrix_xy = gram_matrix_y * K
+
+        P = cvxopt.matrix(gram_matrix_xy.numpy().astype(np.double))
```
Comment (markroxor) [inline fromscratchtoml/models/svm/svc.py]: check this.

File: fromscratchtoml/models/svm/svc.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+        A = cvxopt.matrix(y.numpy().astype(np.double)).trans()
+        b = cvxopt.matrix(0.0)
+
+        lagrange_multipliers = ch.Tensor(list(cvxopt.solvers.qp(P, q, G, h, A,
+                                                                b)['x']))
+
+        lagrange_multiplier_indices = lagrange_multipliers > eta
+
+        self.support_vectors = ch.stack([x for multiplier, x in
```
Comment (markroxor) [inline fromscratchtoml/models/svm/svc.py]: find a better way

File: fromscratchtoml/models/svm/svc.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+                                         zip(lagrange_multipliers, X)
+                                         if multiplier > eta])
+
+        self.support_vectors_y = ch.Tensor(y[lagrange_multiplier_indices])
+        self.lagrange_multipliers = lagrange_multipliers[lagrange_multiplier_indices]
+
+        lagrange_multiplier_indices = lagrange_multiplier_indices.nonzero().view(-1)
+
+        self.b = 0
+        for i in range(self.support_vectors_y.size()[0]):
```
Comment (markroxor) [inline fromscratchtoml/models/svm/svc.py]: `num_supp_vectors`

**Pull request merged**